crates/test/migration-test/tests/network_integration.rs

```diff
@@ -0,0 +1,403 @@
+    thread::sleep(*EPOCH_DURATION);
+}
+
+/// Look up any delegation to a validator, and undelegate it.
+/// Useful for "undoing" the delegate action, which can help with keeping the
+/// test suite somewhat idempotent.
+fn undelegate() {
+    tracing::info!("check that we have some of the delegation token");
+    let tmpdir = load_wallet_into_tmpdir();
+    let validator = get_validator(&tmpdir);
+    // Check we have some of the delegation token for that validator now.
+    let mut balance_cmd = get_pcli_path();
+    balance_cmd
+        .args(["--home", tmpdir.path().to_str().unwrap(), "view", "balance"])
+        .timeout(std::time::Duration::from_secs(TIMEOUT_COMMAND_SECONDS));
+
+    balance_cmd
+        .assert()
+        .stdout(predicate::str::is_match(validator.as_str()).unwrap());
+
+    let balance_output = balance_cmd.output().unwrap().stdout;
+    let balance_output_string = String::from_utf8_lossy(&balance_output);
+
+    tracing::debug!(?balance_output_string, "balance output string");
+
+    // We successfully delegated. But since the validator exchange rates are dynamic, we
+    // need to pull the amount of delegation tokens we obtained so that we can later
+    // try to execute an undelegation (`tx undelegate <AMOUNT><DELEGATION_TOKEN_DENOM>`).
+    // To do this, we use a regex to extract the amount of delegation tokens we obtained:
+    let delegation_token_pattern = Regex::new(r"(\d+\.?\d?m?delegation_[a-zA-Z0-9]*)").unwrap();
+    let (delegation_token_str, [_match]) = delegation_token_pattern
+        .captures(&balance_output_string)
+        .expect("can find delegation token in balance output")
+        .extract();
+
+    tracing::info!("attempting undelegation");
+    let mut undelegate_cmd = get_pcli_path();
+    undelegate_cmd
+        .args([
+            "--home",
+            tmpdir.path().to_str().unwrap(),
+            "tx",
+            "undelegate",
+            delegation_token_str,
+        ])
+        .timeout(std::time::Duration::from_secs(TIMEOUT_COMMAND_SECONDS));
+    undelegate_cmd.assert().success();
+
+    tracing::info!("undelegation succeeded, wait an epoch before claiming.");
+    // Wait for the epoch duration.
+    thread::sleep(*EPOCH_DURATION);
+    tracing::info!("epoch passed, claiming now");
+    let mut undelegate_claim_cmd = get_pcli_path();
+    undelegate_claim_cmd
+        .args([
+            "--home",
+            tmpdir.path().to_str().unwrap(),
+            "tx",
+            "undelegate-claim",
+        ])
+        .timeout(std::time::Duration::from_secs(TIMEOUT_COMMAND_SECONDS));
+    tracing::info!(?undelegate_claim_cmd, "claiming");
+    undelegate_claim_cmd.assert().success();
+    tracing::info!("success!");
+    sync(&tmpdir).expect("sync is successful");
+}
+
+/// TOML for an "upgrade-plan" governance proposal.
+// Intentionally avoiding importing this type to adhere to strict
+// CLI interfacts for the pcli binaries.
```

Comment:
```suggestion
// CLI interfacts for the pcli binaries.
```

i think a sentence got garbled here!